Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/publish.py

```diff
@@ -0,0 +1,251 @@
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import re
+import sys
+
+from .command import Command
+from .track import Track
+from webkitcorepy import arguments, run, string_utils, Terminal
+from webkitscmpy import log, local
+
+
+class Publish(Command):
+    name = 'publish'
+    help = "Push a specific branch or tag and all its history to a specified remote."
+
+    @classmethod
+    def parser(cls, parser, loggers=None):
+        parser.add_argument(
+            'arguments', nargs='+',
+            type=str, default=None,
+            help='String representation(s) of branches or tags to be published',
+        )
+        parser.add_argument(
+            '--remote', dest='remote', type=str, default=None,
+            help='Publish content to the specified remote',
+        )
+
+    @classmethod
+    def branches_on(cls, repository, ref):
+        output = run(
+            [repository.executable(), 'branch', '-a', '--merged', ref],
+            cwd=repository.root_path,
+            capture_output=True,
+            encoding='utf-8',
+        )
+        if output.returncode:
+            sys.stderr.write(output.stderr)
+            return {}
+        result = {None: set()}
+        for line in output.stdout.splitlines():
+            split = line.lstrip().split('/', 2)
+            if len(split) < 3 or split[0] != 'remotes':
+                result[None].add('/'.join(split))
+            else:
+                if split[1] not in result:
+                    result[split[1]] = set()
+                result[split[1]].add(split[2])
+        return result
+
+    @classmethod
+    def tags_on(cls, repository, ref):
+        result = run(
+            [repository.executable(), 'tag', '--merged', ref],
+            cwd=repository.root_path,
+            capture_output=True,
+            encoding='utf-8',
+        )
+        if not result.returncode:
+            return result.stdout.splitlines()
+        sys.stderr.write(result.stderr)
+        return []
+
+    @classmethod
+    def parental_intersection(cls, repository, commit):
+        if commit.branch == repository.default_branch:
+            return None
+        if commit.identifier <= 1:
+            return None
+        try:
+            parent_branch = repository.find('{}~{}'.format(commit.hash, commit.identifier - 1)).branch
+        except (ValueError, repository.Exception):
+            return None
+        if parent_branch == commit.branch:
+            return None
+        remote = repository.remote_for(parent_branch)
+        intersection = repository.merge_base(commit.hash, 'refs/remotes/{}/{}'.format(remote, parent_branch))
+        if intersection.hash == commit.hash:
+            return None
+        return intersection
```

Comment:
Could you help me understand why we need to look up _both_ the commit at ID=1, _and_ the merge-base of `commit` + the parent branch? I would think that the commit at ID=1 would _be_ on the parent branch, making the merge-base call unnecessary.

Reply:
We have 2 questions we need to answer.

The first is "what is my parent branch". That's actually an extremely difficult question to answer, because `git` doesn't have an idea about which branch came first. People tend to have a concept of this, though, and the WebKit project has some reliable patterns we can use. Generally, we have `main` and then long running branches off of `main`, and then specific release branches off of that long running branch. (sometimes there is actually a 3rd branch between the long running one and the release branches, but we aren't consistent about naming conventions in that case, so it's not easy to determine if two release branches are just sharing a few commits or if we actually have one of these 3rd level branches). Ultimately, we usually care about that long-running branch more than the specific release branch, but our `git-webkit publish` command will almost certainly be handed a branch or a tag that corresponds to a specific release branch. We need to find what the long-running branch associated with the release branch we're on, and `identifier=1` will be the first commit on this long-running branch. If folks are following the right conventions with branch naming, we'll end up getting the long-running branch's name as the "canonical" branch name for that first commit diverging from `main`

The second question we need to answer is "how much of the long-running branch should we be publishing". We can't publish all of the long-running branch, because it's likely that the long-running branch has more commits than just the ones on our release branch, so we only want to publish the long-running branch up to the newest merge-base of the refs the user provided, hence the merge-base. You'll notice that the function returns a commit and not a branch, this is the reason why. We aren't necessarily pushing the entirety of every branch.